Source/Core/DolphinQt/Debugger/CodeDiffDialog.cpp

```diff
@@ -0,0 +1,408 @@
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_include = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    current_diff = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+      }
+    }
+  }
+
+  if (!m_include.empty())
+  {
+    // Compare include with current and remove items that aren't in both.
+    std::vector<Diff> tmp_swap;
+    for (auto& iter : m_include)
+    {
+      if (std::any_of(current_diff.begin(), current_diff.end(),
+                      [&](Diff& v) { return v.symbol == iter.symbol; }))
+      {
+        tmp_swap.push_back(iter);
+      }
+    }
+    m_include.swap(tmp_swap);
+  }
+  else
+  {
+    m_include = recorded_symbols;
+    RemoveMatchingSymbolsFromIncludes(m_exclude);
+  }
+}
+
+void CodeDiffDialog::OnExclude()
+{
+  const auto recorded_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+
+        m_exclude.insert(pos, iter);
+      }
+    }
+    // If there is no include list, we're done.
+    if (m_include.empty())
+      return;
+  }
+
+  if (!m_exclude.empty())
+  {
+    RemoveMatchingSymbolsFromIncludes(current_diff);
+  }
+}
+
+std::vector<Diff> CodeDiffDialog::CalculateSymbolsFromProfile()
+{
+  Profiler::ProfileStats prof_stats;
+  auto& blockstats = prof_stats.block_stats;
+  JitInterface::GetProfileResults(&prof_stats);
+  std::vector<Diff> current;
+  current.reserve(20000);
+
+  // Convert blockstats to smaller struct Diff. Exclude repeat functions via symbols.
+  for (auto& iter : blockstats)
+  {
+    Diff tmp_diff;
+    std::string symbol = g_symbolDB.GetDescription(iter.addr);
+    if (!std::any_of(current.begin(), current.end(),
+                     [&symbol](Diff& v) { return v.symbol == symbol; }))
+    {
+      tmp_diff.symbol = symbol;
+      tmp_diff.addr = iter.addr;
+      tmp_diff.hits = iter.run_count;
+      current.push_back(tmp_diff);
+    }
+  }
+
+  // Could add address based difference instead of symbols. Probably need second function.
+  // Sort for lower_bound.
+  sort(current.begin(), current.end(),
+       [](const Diff& v1, const Diff& v2) { return (v1.symbol < v2.symbol); });
+
+  return current;
+}
+
+void CodeDiffDialog::RemoveMatchingSymbolsFromIncludes(const std::vector<Diff>& symbol_list)
+{
+  for (auto& list : symbol_list)

```

Comment:
Can't this be simplified as:
```c++
m_include.erase(std::remove_if(m_include.begin(), m_include.end(), [&](Diff& i) {
  return std::any_of(symbol_list.begin(), symbol_list.end(), [&](Diff& s) {
    i.symbol == s.symbol;
  });
}), m_include.end());
```

Reply:
Is there any performance benefit?  I'm not sure which I prefer..

Reply:
I think there are some perform benefits. Each call to erase will reconstruct the vector by moving everything down on erased items because it has to be contiguous in memory .

A simple example using [Quickbench](https://quick-bench.com/q/M3BxPSz1o3j2v0BqPZvjYV8Bg30)
![image](https://user-images.githubusercontent.com/7890055/103663072-216ffa80-4f8a-11eb-942c-af22fd2075d2.png)

Reply:
Updated to your simplified func

Reply:
@sepalani if I go with your implementation instead of the for, the parameter in the signature must be non-constant:

`void CodeDiffDialog::RemoveMatchingSymbolsFromIncludes(std::vector<Diff>& symbol_list)`

instead of:

`void CodeDiffDialog::RemoveMatchingSymbolsFromIncludes(const std::vector<Diff>& symbol_list)`

algorithm blocks compilation with a C2664 otherwise.

Reply:
My example was using a const reference too. The only non-const reference was the vector I wanted to filter which is `m_include` in our case and is not a parameter. So using `void CodeDiffDialog::RemoveMatchingSymbolsFromIncludes(const std::vector<Diff>& symbol_list)` should be fine.

Maybe I'm missing something obvious, I'll do some testing on my end to make sure. Please, feel free to disregard my review below regarding that const reference issue meanwhile.

Reply:
Just getting back to this PR.
Still getting this.
```
Error	C2664
'bool CodeDiffDialog::RemoveMatchingSymbolsFromIncludes::<lambda_1>::()::<lambda_1>::operator ()(Diff &) const': cannot convert argument 1 from 'const Diff' to 'Diff &'
        if (_Pred(*_UFirst)) {
^ (compiling source file Debugger\CodeDiffDialog.cpp
```


I've modified the .h signature to be:
`void RemoveMatchingSymbolsFromIncludes(const std::vector<Diff>& symbol_list);`

For now making the changes without const signature.

Reply:
Clarifying- modified to match minus passing the param as const. The rest has been changed to match sepalani's more performant code.

Reply:
Update: re-adding const now that I've realized my mistake.

Done.